pkg/network/protocols/events/batch_offsets.go

```diff
@@ -0,0 +1,93 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+// +build linux_bpf
+
+package events
+
+import (
+	"sync"
+)
+
+// offsetManager is responsible for keeping track of which chunks of data we
+// have consumed from each batch object
+type offsetManager struct {
+	mux        sync.Mutex
+	stateByCPU []*cpuReadState
+}
+
+type cpuReadState struct {
+	// this is the nextBatchID we're expecting for a particular CPU core. we use
+	// this when we attempt to retrieve data that hasn't been sent from kernel space
+	// yet because it belongs to an incomplete batch.
+	nextBatchID int
+
+	// information associated to partial batch reads
+	partialBatchID int
+	partialOffset  int
+}
+
+func newOffsetManager(numCPUS int) *offsetManager {
+	stateByCPU := make([]*cpuReadState, numCPUS)
+	for i := range stateByCPU {
+		stateByCPU[i] = new(cpuReadState)
+	}
+
+	return &offsetManager{stateByCPU: stateByCPU}
+}
+
+// Get returns the data offset that hasn't been consumed yet for a given batch
+func (o *offsetManager) Get(cpu int, batch *batch, syncing bool) (i, j int) {
+	o.mux.Lock()
+	defer o.mux.Unlock()
+	state := o.stateByCPU[cpu]
+
+	if batchComplete(batch) {
+		// update nextBatchID
+		state.nextBatchID = max(int(batch.Idx)+1, state.nextBatchID)
```

Comment:
I think I'm missing something here
why would the nextBatchID remain the same?

Reply:
good point. this was meant to address (unlikely) scenarios such as the following:
1) batch-0 is filled up in eBPF
2) Consumer.Sync() is called, which means batch-0 data is fetched
"preemptively" before it is consumed from the perf
buffer. nextBatchID is updated to 1;
3) batch-1 is filled up in eBPF
4) Consumer.Sync() is called again; nextBatchID is updated to 2;
5) Now batch-0 is finally read from the perf buffer
we obviously don't want to set nextBatchID back to 1 which is why
update it the like we do

but now I realize that we should add this check to the beginning of the function as well:
```go
if batchID < state.nextBatchID {
	return 0, 0
}
```
and by doing so we can simply update `nextBatchID` like the following:

```go
state.nextBatchID = int(batch.Idx) + 1
```